### PR TITLE
Feature/158 name registry register application

### LINKED ISSUE #158: [name-registry] Support registering an application
**Add a new method `registerApp(did: string, ownerName: string, appName: string, metadata: string[])`. This registers an application as being owned by a given did and associates a domain name with it.**

For example; Register the Verida: Vault application context to be controlled by a Verida did with metadata:

```
{
  domain: 'verida.network',
  icon: 'https://assets.verida.network/icon/png'
}
```

**Requirements:**

- The `registerApp()` request must be signed by the owner did. Throws invalid request signature.
- A `did` can register multiple applications, but all applications they register must have the same `ownerName`. Throws `did` is associated with a different owner.
- The combination of `ownerName` and `appName` must be unique. Throws application already exists error.
- Both `ownerName` and `appName` must only contain alphanumeric characters (including spaces). Throws invalid owner or application name error.
- `metadata.domain` is required, throws domain is required error
- `metadata.domain` must be all lower-case and match the valid characters for a domain name. Throws invalid domain format error.
- The `did` must provide `$APP_FEE` VDA tokens as a fee for registering an application. These tokens can be withdrawn by the smart contract owner. Throws insufficient tokens provided error.
- `metadata` is an array of key / value pairs

**Add a new method `getApp(ownerName: string, appName: string)`. This returns the `did` and `metadata`.**

**Requirements:**

- If the combination of `ownerName` and `appName` is not found, throws application not found error.

**Add a new method `deregisterApp(did: string, ownerName: string, appName: string)`. This de-registers an application.**

**Requirements:**

- The `deregisterApp()` request must be signed by the owner did. Throws invalid request signature
- If the combination of `ownerName` and `appNam`e is not found, throws application not found error.

**Add a new method `updateApp(did: string, ownerName: string, appName: string, domain: string, metadata: string)`. This updates the metadata for a registered application.**

**Requirements:**

-The `updateApp()` request must be signed by the owner did. Throws invalid request signature
- The same verification rules apply as `registerApp()`.

### PR DESCRIPTION
**Pull request by ITStar10**

### CODE REVIEW

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+        if (!_didApps[did].contains(appName)) {
+            revert AppNotFound(false, true);
+        }
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     */
+    function deRegisterApp(
```
Comment (tahpot) [inline VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: `deregisterApp`

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+        }
+
+        // de-register
+        delete _didOwnerName[did];
+        delete _ownerNameInfo[_ownerName];
+
+        _didApps[did].remove(_appName);
+        delete _didAppMetaData[did][_appName];
+
+        emit DeRegisterApp(did, _ownerName, _appName);
```
Comment (tahpot) [inline VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: `emit deregisterApp`

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+     * @dev See {INameRegistryV2}
+     */
+    function getAppRegisterFee() external view virtual override returns(uint) {
+        return appRegisterFee;
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     */
+    function enableAppRegister(bool isEnabled) external virtual payable override onlyOwner {
```
Comment (tahpot) [inline VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Need `disableAppRegister`

**Review by tahpot — APPROVED** [review]
Approved with minor cahnges

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 1 above)
Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Updated

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 2 above)
Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: All the other event names are started with the uppercase letter.
For example, the event for the `registerApp()` function is `RegisterApp`.
How about to name the event for the `deregisterApp()` function as `Deregisterapp`?

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 3 above)
Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: As we've already deployed the `NameRegistry` contract on the testnet and mainnet, we should upgrade the current contracts. Here, the registering app feature is now disabled in the original `NameRegistry` contract.
After we upgrade the contract, we should set the fee for registering app and set the token address. 
And then, we should enable the app registering by call this `enableAppRegister(true)` function.

So, I think the function name of `enableAppRegister()` might be suitable.

Comment (tahpot) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: I agree with the name, I'm saying we should have an extra method to disable app registration. We may want to do that in the future.

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 2 above)
Comment (tahpot) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Yes `DeregisterApp`

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 3 above)
Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Once we called `enableAppRegister(false)`, it would disable the registering app.

Comment (tahpot) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Oh I see, sorry I missed the boolean.

In that case, make it `isAppRegisterEnabled()`

Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: There is already `isAppRegisterEnabled()` function that returns the boolean.
image

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 2 above)
Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Updated

File: VDA-Name-Registry/contracts/v2/NameRegistryV2.sol (hunk 3 above)
Comment (tahpot) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Haha ok.

Let's call it `setAppRegisterEnabled()`

Comment (ITStar10) [reply VDA-Name-Registry/contracts/v2/NameRegistryV2.sol]: Updated

**Review by tahpot — APPROVED** [review]

**Pull request merged**